src/Tasks/AssemblyDependency/ResolveAssemblyReference.cs

```diff
@@ -986,14 +987,23 @@ public ITaskItem[] SuggestedRedirects
         /// </summary>
         private List<ITaskItem> _filesWritten = new List<ITaskItem>();
 
+        private ITaskItem[] _filesWrittenArray = null;
+
         /// <summary>
         /// The names of all files written to disk.
         /// </summary>
         [Output]
         public ITaskItem[] FilesWritten
         {
             set { /*Do Nothing, Inputs not Allowed*/ }
-            get { return _filesWritten.ToArray(); }
+            get
+            {
+                if (_filesWrittenArray?.Length != _filesWritten.Count)
```

Comment:
This condition surprises me. Is a length change the only way it can be invalidated?

What is the reason to do this caching rather than just do it? This field will be read once on task completion in the normal case, right?

Reply:
I originally had it without the caching but was told I should change it. See https://github.com/dotnet/msbuild/pull/6533#discussion_r646947771

Having a length check was probably a bad plan. It would be quite expensive to do a proper check, though, so I think I should just remove the caching here.

Reply:
Yeah, let's remove it, unless there are in-code accesses to these getters. The engine itself should access them only once/task invocation.